SPI fixes

mbolivar opened the pull request:
This series makes improvements to the driver core for, and fixes bugs in, new-style SPI drivers.

- Add status reporting to the SPI driver core's synchronous calls to enable better error handling and reporting
- Add SPI support to 96b_carbon (STM32F4 based)
- Make various fixes to the STM32 SPI driver

Without this series, SPI master on STM32F4 is not working in my testing. With it, the loopback tests pass on 96b_carbon, and I've also verified that the behavior of other protocol features (like NSS pins) not tested in loopback are working correctly using a logic analyzer.

I don't however have an STM32F3 or other board with the FIFO-based IP block to test on -- @erwango, @superna9999, @ldts, any testing here would be much appreciated.

Review requested from @tbursztyka as well.

Commit:
pinmux: stm32: 96b_carbon: Add support for SPI

Add SPI pin mux tables for 96b_carbon. Note that SPI1 is for
internal use on the board connecting to the 96b_carbon_nrf51, while
SPI2 is broken out to an expansion header.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
drivers: spi: stm32: fix hardware NSS management

The LL_SPI_NSS_* macros used in spi_stm32_configure() when
hardware-based NSS management is requested are incorrect; fix them.

In master mode, this seems like a copy/paste error. The slave mode
case is likely due to following incorrect documentation in the ST LL
headers.

Note that in my testing on STM32F4, NSS appears to be open drain when
managed by hardware, making that configuration harder to test (and
probably less useful).

Details for the curious:

    The ST LL headers (for example stm32f4xx_ll_spi.h) claim
    LL_SPI_NSS_HARD_INPUT is to be used only in master mode, and
    LL_SPI_NSS_HARD_OUTPUT is to be used in slave mode.

    The opposite is true: when NSS is not handled by software, the SPI
    peripheral is responsible for driving NSS as an output, and the
    slave peripheral is responsible for reading it as an input.

    This is an error in the LL header files; the reference manuals and
    the other LL code make this clear.

    - The ST reference manuals specify that LL_SPI_HARD_OUTPUT (which
      corresponds to SSM unset, SSOE set) is a master-only
      configuration.  For example, STM32 RM0368 says:

        "NSS output enabled (SSM = 0, SSOE = 1)

        This configuration is used only when the device operates in
        master mode."

    - LL_SPI_HARD_INPUT (SSM unset, SSOE unset) is either a master or
      a slave configuration; in the slave case (which is what we're
      interested in here), it corresponds to the "usual" NSS
      input. RM0368, again:

        "NSS output disabled (SSM = 0, SSOE = 0)

        This configuration allows multimaster capability for devices
        operating in master mode. For devices set as slave, the NSS
        pin acts as a classical NSS input: the slave is selected when
        NSS is low and deselected when NSS high."

      The LL_SPI_StructInit() implementations similarly combine
      LL_SPI_MODE_SLAVE with LL_SPI_NSS_HARD_INPUT.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
drivers: spi: avoid undefined behavior

Void pointer arithmetic is undefined behavior (UB).

It's OK for struct spi_buf to contain a void *, because those values
are only ever stored, read, and compared. However, pointer arithmetic
is done on the tx_buf and rx_buf fields in struct spi_context, so
those need to be u8_t * to avoid UB.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
drivers: spi: stm32: remove extraneous unaligned macros

Byte access is always naturally aligned; there's no need to use
UNALIGNED_GET or UNALIGNED_PUT. Those would only be needed when
supporting 16-bit data frames.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
drivers: spi: report synchronous completion status

The SPI API allows waiters to block until SPI I/O has completed. The
asynchronous subset of the API allows waiters to learn a status value
for the result of the I/O. However, the synchronous API does not allow
this.

Due to this limitation, synchronous API users cannot learn when
interrupt-driven I/O fails, which precludes proper error handling.

Resolve this limitation by adding a sync_status field to struct
spi_context, and using it to return operation results to the waiter.

Since there is only one status field, reduce the maximum number of
supported waiters from UINT_MAX to 1. This is not a problem for
current users, which all wait with the entire context locked.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
drivers: spi: stm32: add error checking to IRQ mode

Now that struct spi_context supports passing errors from
interrupt-driven I/O handlers to waiting threads, we can enable error
interrupts and propagate errors to spi_transceive() callers.

To make it easier for users to debug SPI-related issues, log any error
bits set in SR when failures occur.

A subsequent patch will add error checking to polled mode as well, but
other cleanups and fixes will go in first to make this easier.

Note that this breaks the spi_loopback test on some targets, but it's
not a regression, as it wasn't working properly anyway. Subsequent
patches the bugs that this error checking has exposed.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
drivers: spi: stm32: fix bugs in polled mode

In polled mode, the STM32 SPI driver is signaling completion when
there are no waiters:

- the only spi_context_wait_for_completion() caller in this driver is
  in the IRQ-driven portion of transceive() itself, which isn't
  compiled in polled mode.

- the "asynchronous completion + polled I/O" combination is not
  supported by the driver, so there are no other threads polling on
  this I/O we need to signal completion to.

What should be happening instead of signaling completion is releasing
the chip select pin, which polled I/O currently doesn't do.

Fix these issues.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
drivers: spi: stm32: factor out completion routine

Polled and IRQ-driven SPI I/O share code for cleanup and completion,
which can now be factored into its own routine.

This keeps a single point of truth for common paths, which will allow
a subsequent bug fix to happen in one place, and help avoid future
regressions.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
drivers: spi: stm32: fix SPI being left on in slave mode

The current implementation unconditionally enables the SPI (sets
SPI_CR1_SPE) in transceive(), but disables it only in master mode.

The peripheral should only be enabled while the user has specifically
requested I/O. Fix this by always disabling the peripheral when I/O is
complete.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
drivers: spi: stm32: add error checking to polled mode

With some other issues in polled mode now resolved, add error handling
and report a valid error status when releasing the context.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
tests: spi_loopback: fix typos

Trivial fixups.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
tests: spi_loopback: log start/end of test at INF level

This makes it easier to distinguish which lines come from which tests.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
tests: spi_loopback: add 96b_carbon support

This includes files for both polled and IRQ-driven I/O. The log level
is deliberately kept to ERROR to try to shake out any timing-dependent
bugs which go away with extra logging.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
tests: spi_loopback: improve error logging

Print the entire memory areas of interest on error, instead of
stopping at the first null character.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
drivers: spi: fix typo

There's an extraneous "x" character in a comment.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
drivers: spi: add SPI_OP_MODE_GET()

This is a convenience macro for getting the master/slave operational
mode, which will be used in a subsequent commit.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
drivers: spi: stm32: fix transmit/receive procedure

The transmit and receive procedure used in the STM32 SPI driver is not
correct.

On STM32F4, this is causing OVR errors (per the logged error mask) and
transmission of undesired 0x00 bytes (verified with a logic analyzer).

The root cause is that the receive register is not read (via DR, when
RXNE is set) each time the transmit register is written (also via DR,
when TXE is set). This clearly causes OVR errors when there is no
FIFO, as the receive register needs to be read each time a frame is
transceived, or the IP block has no way of knowing that the
overwritten data were not important.

Adapt the I/O procedure so that every DR write is matched by a DR
read, blocking until the relevant flags are set if necessary.

This behavior is suboptimal for targets such as STM32L4, where there
is a SPI FIFO. However, SPI I/O is broken on those targets, and this
patch fixes them as well. Further optimizations for targets with FIFOs
is left to future work.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

erwango commented:
Hi @mbolivar,
I had some quick tests. not much time to analyze more for now:
*nucleo_l476rg: test is stalled (irq & poll)
   > [general] [INF] spi_complete_loop: Start ..... Then hangs
   > in spi_stm32_shift_m, waiting for RXNE flag
   > regression vs previous status
*nucleo_f334r8: bus error: (irq & poll)
  Faulting instruction address:  0x8001f0e  (drivers/spi/spi_context.h:229)
  Imprecise data bus error
Fatal fault in ISR! Spinning...
   > Not a regression (hum... this is embarrassing :/)

Hope it helps.

mbolivar commented:
> I had some quick tests. not much time to analyze more for now:

Awesome, thanks for the testing!

So something is wrong with this implementation in the case of SPI peripherals with FIFOs (F3 and L4). It looks like we'll need to select a spi_stm32_shift_frame() implementation depending on the STM32 family's SPI peripheral type.

It looks like this will be easy to do if we check for the presence of the FIFO threshold bit in CR2:

```
#ifdef SPI_CR2_FRXTH
/* FIFO spi_stm32_shift_frame() implementation goes here */
#else
/* non-FIFO spi_stm32_shift_frame() implementation goes here */
#endif
```

Seem OK and portable across STM32 SoCs?

(I would really prefer to avoid having to add in a CONFIG_SPI_STM32_VERSION that we then fill into the Kconfig for every SoC, or, worse, board defconfig when the information we need is available to us from CMSIS already. I.e. take a different approach than the STM32 I2C "V1" and "V2" configuration mechanism.)

mbolivar commented:
@erwango I've rebased and updated this per our discussion in email. I've tried to keep the existing behavior unchanged in the FIFO case. If you have some time to retest, I'd appreciate it. Tests are passing on STM32F4.

erwango commented:
@mbolivar, test goes a little bit further on L476:
`�[general] [INF] main: SPI test on buffers TX/RX 0x2000000c/0x20000274
[general] [DBG] spi_async_call_cb: Polling...
[general] [INF] spi_complete_loop: Start
[general] [DBG] spi_stm32_configure: Installed config 0x20000038: freq 312500Hz (div 0
[general] [DBG] spi_context_buffers_setup: tx_bufs 0x20000b50 (1) - rx_bufs 0x20000b51
[general] [DBG] spi_context_buffers_setup: current_tx 0x20000b50 (1), current_rx 0x207
[general] [DBG] spi_context_update_tx: tx buf/len 0x2000000d/16`
But stall and remains stuck at:
`0x08001d00 in spi_stm32_shift_m (data=0x200000c0 <spi_stm32_dev_data_1>, spi=0x40013000) at /local/mcu/zephyr/zephyr-project/drivers/spi/spi_ll_stm32.c:68
68		while (!LL_SPI_IsActiveFlag_RXNE(spi)) {`

F334r8 has same status as previously. For this one, since we know this is not a regression and
this is not linked with your dev, I'll raise a JIRA (or github issue, or..)

mbolivar commented:
> But stall and remains stuck at:

Hrm, I wasn't expecting that :(.

I'll wait to get some L4 hardware to test on here and update this PR when I've had a chance to look into this further.

mbolivar commented:
Thanks to @erwango 's kindly sending me an L4-based board, I was able to retest this on that SPI IP. I've rebased and updated the series to include some fixes for STM32L4 as well, which is also currently broken on mainline, and the problems with which I'm grateful to Erwan for diagnosing.

I've verified spi_loopback functionality with this series on STM32L4 and STM32F4 in both polled and loopback modes.

@tbursztyka @erwango -- could you please review?

Commit:
CODEOWNERS: update drivers/spi/spi_ll_stm32.*

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

mbolivar commented:
Not sure why shippable is failing after I added the CODEOWNERS commit; any ideas? Looks like a flaky test.

ydamigos commented:
@mbolivar I got the same error yesterday. Anas informed me that it wasn't my fault.

erwango reviewed (approved):
LGTM
As mentioned in comments:
*Tested ok on L4
*Issue on F3 which does not look related to this fix. ZEP-2419 raised for this purpose.

mbolivar commented:
Thanks, Erwan.

@tbursztyka, I'm especially hoping for your OK on de50763, which affects the driver core.

mbolivar commented:
Rebased and dropped the first two patches from the previous version, which were made obsolete by #888. Tests still pass on F4 and L4.

superna9999 commented:
Sorry I was in vacation, I'll get a run of the patchset.

First of all, I don't really understand why you treat master and slave shifting differently.
From a pure SPI bus PoV, it's totally identical, did you run into an issue somewhere ?

superna9999 commented on CODEOWNERS:
```diff
@@ -73,6 +73,7 @@ drivers/serial/uart_altera_jtag.c        @andrewboie
 drivers/serial/uart_riscv_qemu.c         @fractalclone
 drivers/slip/*                           @jukkar @tbursztyka
 drivers/spi/*                            @tbursztyka
+drivers/spi/spi_ll_stm32.*               @mbolivar
```
please add the other contributors like @ldts and me

mbolivar replied:
Do you want to add yourselves? There's no issue in doing so -- I just didn't want to "volunteer" you for future review.

superna9999 commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -25,52 +25,152 @@
+/* Value to shift out when no application data needs transmitting. */
+#define SPI_STM32_TX_NOP 0x00
+
+static bool spi_stm32_transfer_ongoing(struct spi_stm32_data *data)
```
Maybe you can force it as inline

mbolivar replied:
I want to let GCC decide. Any issues?

superna9999 commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -25,52 +25,152 @@
+	return spi_context_tx_on(&data->ctx) || spi_context_rx_on(&data->ctx);
+}
+
+static int spi_stm32_get_err(SPI_TypeDef *spi)
```
Same here

mbolivar replied:
Same; I like to let the compiler decide.

superna9999 commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -25,52 +25,152 @@
+		*data->ctx.rx_buf = rx_frame;
+		spi_context_update_rx(&data->ctx, 1);
+	}
+}
```
I don't get the true difference apart that in the master shift you consider you want to push a byte and then get a byte, but why is it different for slave and why can't we leave the previous way for both ?

mbolivar replied:
Since the master is in charge of the bus and controls when frames get shifted, it's able to block until the TX/RX registers are empty/full respectively to move the I/O along. But doing that from the slave felt like asking for trouble, so I made it more of a "best effort" basis.

If by "previous way" you mean the code as is, I tried to explain in my commit logs why it's not correct: to transmit, it writes to DR, then doesn't always read back the frame that got shifted in by waiting until DR has data to read and pulling it out, freeing up the RX register for the next time a frame needs to be clocked out. This is causing overrun errors (OVR bit in SR).

mbolivar commented:
@superna9999 are you OK with my answers to your questions?

superna9999 commented:
@mbolivar Yes, please add me to the code owners, for the other questions, I'm still doubious since it's more an issue on non-FIFO devices, but it will keep code simple, I will live fine with it.

So LGTM

mbolivar commented:
@superna9999 I totally agree with you that the behavior in this series is suboptimal for the FIFO case. My goal is just to have a starting point that works on all targets, since the current driver has issues on some STM32s. Thanks for review; I've added you first to the CODEOWNERS for this driver and rebased.

pfalcon commented:
Anything precluding this from being merged? @tbursztyka , @nashif , can you please have a look?

mbolivar commented:
@nashif @galak given that this has the OK from @erwango and @superna9999 I wonder if you could take a look at the change to the SPI core especially in https://github.com/zephyrproject-rtos/zephyr/pull/832/commits/deb3828e9435bf0a3d0f4833521b59815ef06c2a, to see if this is OK to merge.
